[DP-60] Create custom login UI using AWS cognito library

chanmioh opened the pull request:
CHANGES:
- Added a login button at the top right of the website; clicking this will bring the user to the login page at `/login`
- On the login page, entering login credentials and submitting login will attempt a user authentication. If login is successful, the React context holding info about the user is updated accordingly. 
- With this context, the user's JWT token can be accessed + other authentication info if needed.
- If the user is logged in, the login button of the website will function as a Logout.

- Login UI: <img width="564" alt="image" src="https://user-images.githubusercontent.com/40529597/174089948-af8e76a2-fbe9-4625-9b17-b0a0e2fbb94f.png">
- What user sees while logged in: <img width="199" alt="image" src="https://user-images.githubusercontent.com/40529597/174090423-52f12738-e42a-4109-9271-7feac9a03fde.png">

loafofpiecrust commented on website/src/auth.tsx:
```diff
@@ -2,52 +2,71 @@ import { HubCallback } from "@aws-amplify/core"
-const UserContext = createContext<CognitoUser | null>(null)
+import {
+  CognitoUser,
+} from 'amazon-cognito-identity-js'
```
Right now, I think this is a transitive dependency from the amplify library. Could you run `yarn add amazon-cognito-identity-js`?

loafofpiecrust replied:
A transitive dependency is when a package is installed through another package. So if we are `dailp` and our `package.json` says "we depend on `xxx`" and `xxx` depends on `yyy`, then `yyy` is a transitive dependency of `dailp`.

loafofpiecrust commented on website/src/auth.tsx:
```diff
@@ -2,52 +2,71 @@ import { HubCallback } from "@aws-amplify/core"
+  setUser: (value: CognitoUser | null) => void;
+};
+
+export const UserContext = createContext<UserContextType>({} as UserContextType);
```
General convention is not to export context objects directly (only `XProvider` and `useX`). So I'd say either all the code that needs to call `setUser` goes into one file, or exposing `setUser` via the hook (or a new hook). I lean toward the former because generally new code beyond this shouldn't need to call `setUser`.

chanmioh replied:
Done (I think) I made sure useContext() doesn't get called outside of the auth.tsx file, by creating new hooks. I also moved the login and logout functions to the UserProvider component.

loafofpiecrust commented on website/src/client/layout.tsx:
```diff
@@ -1,16 +1,17 @@
-  const token = creds?.signInUserSession.idToken.jwtToken ?? null
+export const LayoutClient = (p: { children: any }) => {
+  // const creds = useCredentials()
+  const token = useCredentials() ?? null; // uses credentials
```
I would move this `?? null` to inside `useCredentials` so that callsites don't need it.

chanmioh replied:
Thanks, made the change

loafofpiecrust commented on website/src/layout.tsx:
```diff
@@ -18,31 +20,34 @@ import "./wordpress.css"
-            <span className={css.subHeader}>
-              Digital Archive of Indigenous Language Persistence
-            </span>
+      <LayoutClient>
```
Hmm I do think we need `LayoutClient` at this scope but let's move the `Helmet` usage before it since that doesn't call graphql.

```html
<Helmet ...>
</Helmet>

<LayoutClient>
...
```

chanmioh replied:
Change has been made!

loafofpiecrust commented on website/src/auth.tsx:
```diff
@@ -2,52 +2,71 @@ import { HubCallback } from "@aws-amplify/core"
-  }, [])
+  const [user, setUser] = useState<CognitoUser | null>(null);
+
+  // React.useEffect(() => {
```
We definitely need something here to restore remembered logins, the cognito library docs have an example of it (use case 16).

chanmioh replied:
I will be taking a look at the docs for this and let you know if I have questions.

loafofpiecrust commented on website/src/client/signin.tsx:
```diff
@@ -2,19 +2,19 @@ import { Auth } from "aws-amplify"
-    return <Link href="/signin">Sign In</Link>
-  }
-}
+// const ClientSignIn = () => {
```
If you think these components are not needed anymore, you can delete the file.

chanmioh replied:
Aaand it's gone!

loafofpiecrust commented on website/src/pages/login.page.tsx:
```diff
@@ -0,0 +1,146 @@
+              e.preventDefault();
+              context.user?.signOut();
+              context.setUser(null);
+            }}>Logout</Link>
+          : <Link href="/login">Login</Link>}
```
Let's do "Log out" and "Log in" (two words)

chanmioh replied:
Updated!

loafofpiecrust commented on website/src/pages/signup.page.tsx:
```diff
@@ -0,0 +1,61 @@
+import { centeredForm, loginFormBox, positionButton, skinnyWidth, submitButton } from "./login.css";
+
+// TODO: this page is  incomplete
+const SignupPage = () => {
```
FYI you and I hadn't talked about this together yet but we are planning on having more like an "Apply to join" flow, but that might manifest as allowing registration for a very limited privilege account until it's approved to contribute. So the registration page will need further thought but for now let's use `register` as the path to avoid "Sign in/out/up" entirely.

chanmioh replied:
Got it, updated the path.

loafofpiecrust commented on website/src/pages/reset-password.page.tsx:
```diff
@@ -0,0 +1,40 @@
+    const [email, setEmail] = useState("");
+
+    return (<Layout>
+        <main className={skinnyWidth}>
```
Mind running `prettier` on the files you've changed? I use the format-on-save functionality in my editor so I never have to worry about it.

loafofpiecrust reviewed (commented):
Overall looking great and thanks for looking ahead to the registration and reset password flows!

loafofpiecrust commented on website/src/pages/login.page.tsx:
```diff
@@ -0,0 +1,146 @@
+        </div>
+
+        <div className={positionButton}>
+          <input className={submitButton} type="submit" value="Login" />
```
Same here with using two words, I think "login" is the noun and "log in" is the verb.

loafofpiecrust commented on website/src/pages/login.page.tsx:
```diff
@@ -0,0 +1,146 @@
+
+      <form className={centeredForm} onSubmit={e => submitLogin(e, email, password)}>
+        <div >
+          <label>
+            Email *</label>
+          <input
```
Make sure to add `name` to labels and inputs so that they are semantically linked.

chanmioh replied:
Updated!

loafofpiecrust commented on website/src/pages/login.page.tsx:
```diff
@@ -0,0 +1,146 @@
+  // gets the user and its setUser from the context, to update info on the current user
+  const context = useContext(UserContext);
+
+  function submitLogin(e: FormEvent<HTMLFormElement>, username: string, password: string) {
```
I would be inclined to move this function up to the provider and have it provide a simple `logIn` function that takes username and password, which you call here. That way, all code that calls the cognito API lives in a single place.

chanmioh replied:
I thought I'd just comment these quick changes I made to make sure I understood: I added login and logout functions to the context type definition, like so: 
<img width="609" alt="image" src="https://user-images.githubusercontent.com/40529597/174151403-d243c661-666d-4b40-ab92-ddf96f4db5f6.png">

Those functions are defined in the UserProvider, and then I created hooks to access those functions: 
<img width="510" alt="image" src="https://user-images.githubusercontent.com/40529597/174151450-0b9e6358-8d62-4a91-a6c8-e40237579a4d.png">

loafofpiecrust replied:
Hmm I think the `login` and `logout` functions don't need the first `event` argument since they could be called from any kind of event generally.

chanmioh replied:
Removed that event argument!

chanmioh commented:
Pushed a new commit with the following:

CHANGES:
- Moved Login and Logout functions from `login.page.tsx` to `auth.tsx`, and integrated them into a single useUser() hook.
- Added a popover when selecting "Log out" to avoid accidental misclicks.
<img width="274" alt="image" src="https://user-images.githubusercontent.com/40529597/174404344-b055e3ed-cc0b-4085-ac61-08966d5ecdcf.png">

- Replaced old HTML form with Reakit form.
- Added a barebones reset password flow, which can be access by clicking "Reset password" below the login form, or going to `/reset-password`
<img width="552" alt="image" src="https://user-images.githubusercontent.com/40529597/174404046-27e7a1fd-2c20-4109-b1bb-bbda6c3f3de3.png">

User will be prompted to enter a valid email address used to register for an account:
<img width="546" alt="image" src="https://user-images.githubusercontent.com/40529597/174404106-71a2fffe-07e9-4f03-8a59-55ba9b23b660.png">

Then, theoretically they should receive a verification code in their inbox that will allow them to complete this change password form. However, I have not been able to test yet as I cannot access the test account's emails.
<img width="527" alt="image" src="https://user-images.githubusercontent.com/40529597/174404177-05a22f6c-eed5-4755-b96a-93a28c835685.png">

- Removed the incomplete sign up flow (for now) as it is still wip.

loafofpiecrust commented on package.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "dependencies": {
+    "amazon-cognito-identity-js": "^5.2.9"
+  }
```
Ah sorry I wasn't specific! I meant for you to run this under `website` only, not the top-level project. We want this to get added to `website/package.json`

loafofpiecrust commented on website/src/auth.tsx:
```diff
@@ -1,53 +1,132 @@
+    })
+  }
+
+  function logoutUser(user: CognitoUser) {
+    user?.signOut()
```
Would it make sense to just use the `user` object already in scope instead of needing a function argument?

chanmioh replied:
That does make sense - let me go and change that.

loafofpiecrust commented on website/src/pages/login.css.ts:
```diff
@@ -0,0 +1,40 @@
+export const skinnyWidth = style([
+  edgePadded,
+  centeredColumn,
+  { margin: "auto", width: "35%" },
```
Check this out on mobile, it's super skinny. You should check out the `fullWidth` style we use elsewhere which will make an element full width on mobile but only a max pixel size on desktop. I'd start with that.

chanmioh replied:
Got it, modified some things so on mobile it shouldn't look as wonky.

loafofpiecrust commented on website/src/pages/login.css.ts:
```diff
@@ -0,0 +1,40 @@
+  flexDirection: "column",
+  backgroundColor: "white",
+  padding: "16px",
+  borderRadius: "10%",
```
Avoid percentages for this, check the `sprinkles.css.ts` file for some value presets we have. In there, we have `vspace` for vertical padding/margins, `hspace` for horizontal padding/margins, `radii` for `borderRadius` values, `colors` for `color` or `backgroundColor`, etc. I'd replace many of the values you have here with some of those. i.e. `borderRadius: radii.medium` perhaps. Feel free to add more values to those objects as you see fit though!

loafofpiecrust commented on website/src/pages/login.css.ts:
```diff
@@ -0,0 +1,40 @@
+export const positionButton = style({
+  display: "flex",
+  justifyContent: "flex-end",
+  marginTop: "5%",
```
Very uncommon to use percentages for margins and padding. Use `vspace` or `hspace` values like I suggest above.

loafofpiecrust commented on website/src/pages/login.page.tsx:
```diff
@@ -0,0 +1,176 @@
+  const loginForm = useFormState({
+    values: { username: "", password: "" },
+    onValidate: (values) => {
+      if (!values.username || !values.password) {
+        throw (
+          (!values.username && { username: "A username is required" }) || {
+            password: "A password is required",
+          }
+        )
+      }
```
I think this could just be

```ts
if (!values.username) {
  throw ...
} else if (!values.password) {
  throw ...
}
```

Or perhaps something like this, to show both messages if both are missing:

```ts
if (!values.username || !values.password) {
  throw { username: !values.username && "...", password: !values.password && "..." }
}
``

chanmioh replied:
Modified!

loafofpiecrust commented on website/src/pages/login.page.tsx:
```diff
@@ -0,0 +1,176 @@
+      <Popover {...popover} tabIndex={0}>
+        <div className={logoutPopover}>
+          <div>Log out of DAILP?</div>
+          <Link
+            href=""
```
Hmm we have a `Button` component and you should be able to override the styles if that's what you wanted here. I don't remember if we have a pre-built plainer button that is just text. We shouldn't use a `Link` for this.

chanmioh replied:
Done, went ahead and used a div styled with the cleanButton instead!

loafofpiecrust replied:
If it's semantically a button then the html element should always be `button`. Turning div into a button is also an antipattern unless there's a specific reason.

loafofpiecrust reviewed (commented):
Adding some more comments for now. Would you be able to split the `reset-password` stuff into a separate branch so that we could move forward with the `login` page before finishing `reset-password`?

chanmioh commented:
> Adding some more comments for now. Would you be able to split the `reset-password` stuff into a separate branch so that we could move forward with the `login` page before finishing `reset-password`?

Definitely, and I can go ahead and upload the final code I have for this branch to review.

loafofpiecrust commented on website/src/auth.tsx:
```diff
@@ -1,53 +1,95 @@
-      }
+  // refreshes last authenticated user
+  useEffect(() => {
+    if (user != null) {
+      user.getSession(function (err: Error, result: CognitoUserSession | null) {
```
I'm confused, why are you calling this function then doing nothing with the result? Just to log when you're automatically signed in?

chanmioh replied:
While CognitoPool.getCurrentUser() gets the last authenticated user from the userpool, if the page is refreshed, the localStorage is cleared along with the user's tokens, so I added a UseEffect on component mount that will refresh the user's tokens (and restore them to the local storage). The refresh of the tokens is done automatically when the CognitoUser.getSession() method is called.

chanmioh commented on website/src/pages/login.css.ts:
```diff
@@ -0,0 +1,65 @@
+    margin: "auto",
+    "@media": {
+      [mediaQueries.medium]: {
+        width: "35vw",
```
That makes a lot of sense, thank you!
